python/ray/data/_internal/planner/plan_download_op.py

```diff
@@ -181,37 +181,43 @@ def download_bytes_threaded(
         if len(uris) == 0:
             continue
 
-        paths, fs = _resolve_paths_and_filesystem(uris)
-        fs = RetryingPyFileSystem.wrap(
-            fs, retryable_errors=data_context.retried_io_errors
-        )
+        def load_uri_bytes(uri_iterator):
+            """Resolve filesystem and download bytes for each URI.
 
-        def load_uri_bytes(uri_path_iterator):
-            """Function that takes an iterator of URI paths and yields downloaded bytes for each."""
-            for uri_path in uri_path_iterator:
+            Takes an iterator of URIs and yields bytes for each.
+            """
+            for uri in uri_iterator:
                 read_bytes = None
                 try:
+                    # Resolve filesystem for this specific URI
+                    resolved_paths, fs = _resolve_paths_and_filesystem(uri)
+                    fs = RetryingPyFileSystem.wrap(
+                        fs, retryable_errors=data_context.retried_io_errors
+                    )
+                    resolved_path = resolved_paths[0]
```

Comment:
Nit: Is `resolved_paths` always guaranteed to be length-one? If that's an assumption we're making, let's make that assumption explicit with an assertion?